### PR TITLE
Data Generator

### PR DESCRIPTION
**Pull request by BradReesWork**
The generator mimics the OGB mag format. 

execution example:

python gnn_cite_graph_gen.py \
    -mg True\
    -outdir '/tmp/GNN2' \
     -papersScale 12  \
     -format "csv" \
     -papersLabeledPercent 0.10 \
     -papersFeatureNoise 0.0001 

these uses default of 10 features (-papersNumFeatures)

### CODE REVIEW

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py [hunk 1]
```diff
@@ -0,0 +1,639 @@
...
+    parser.add_argument(
+        "-papersScale",
+        type=int,
+        default=None,
+        help="Scale factor of the Papers-Cite_Papers graph",
+        required=True,
+    )
+
+    parser.add_argument(
+        "-papersEgefactor",
```
Comment (alexbarghi-nv) [inline benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:511]: typo in "EdgeFactor"

**Comment by alexbarghi-nv** [comment]
@BradReesWork I keep getting this error: Size of output exceeds column size limit

Maybe we need to convert to host before writing CSV data?

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+
+def save_data(args, dir, file_name, data, mg=False):
+
+    if mg:
+        f = os.path.join(dir, f"{file_name}")
+    else:
+        f = os.path.join(dir, f"{file_name}.{args.format}")
+    print(f"\tsaving to {f}")
+
+    if format == "parquet":
```
Comment (alexbarghi-nv) [inline benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:137]:
```suggestion
    if args.format == "parquet":
```

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py [hunk 3]
```diff
@@ -0,0 +1,639 @@
...
+
+    # selected just the labeld nodes
+    labeled_df = (nodes["label"].loc[nodes["label"] == 1]).to_frame().reset_index()
+
+    # drop the "labeled" column since it is no longer needed
+    labeled_df.drop(columns=["label"], inplace=True)
+
+    # Save the list
+    dir = args.outdir + paper_dir
+    name = "node-label." + args.format
```
Comment (alexbarghi-nv) [inline benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:283]:
```suggestion
    name = "node-label"
```

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py [hunk 4]
```diff
@@ -0,0 +1,639 @@
...
+    save_data(args, dir, name, labeled_df)
+
+    # some cleanup
+    del ran
+    del labeled_df
+    nodes.drop(columns=["rand"], inplace=True)
```
Comment (alexbarghi-nv) [inline benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:289]:
```suggestion
    nodes.drop(columns=["rand", "label"], inplace=True)
```

**Comment by alexbarghi-nv** [comment]
Could you please also output a JSON containing the number of nodes of each type and the starting offsets?

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py (hunk 2 above)
Comment (BradReesWork) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:137]: oops.  I changed the input argument and missed this one

**Review by VibhuJawa — COMMENTED** [review]
Looks good to me . Thanks for working on this Brad

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py (hunk 3 above)
Comment (BradReesWork) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:283]: in next push

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py (hunk 4 above)
Comment (BradReesWork) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:289]: in next push

File: benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py (hunk 1 above)
Comment (BradReesWork) [reply benchmarks/shared/python/data_generators/gnn_cite_graph_gen.py:511]: oops, fixed

**Comment by BradReesWork** [comment]
> 
You need to save a parquet in that case.  Seems like the CSV writer is arrow based, so no more than 2B rows

**Comment by BradReesWork** [comment]
"Could you please also output a JSON containing the number of nodes of each type and the starting offsets?"

that can be done.  currently there is no starting offset since all vertex IDs are contiguous.   The function accept a starting offset as input, that could be 0 (zero) and an offset JSON file produced

**Comment by alexbarghi-nv** [comment]
By offsets, I mean the starting vertex ID for each vertex type

**Comment by alexbarghi-nv** [comment]
preempted by #3223

**Pull request closed without merging**